python/ray/data/_internal/datasource/parquet_datasink.py

```diff
@@ -109,71 +100,39 @@ def write_blocks_to_path():
             max_backoff_s=WRITE_FILE_RETRY_MAX_BACKOFF_SECONDS,
         )
 
-    def _write_single_file(
+    def _write_parquet_files(
         self,
-        path: str,
         tables: List["pyarrow.Table"],
         filename: str,
         output_schema: "pyarrow.Schema",
         write_kwargs: Dict[str, Any],
     ) -> None:
-        import pyarrow.parquet as pq
+        import pyarrow.dataset as ds
 
-        # We extract 'row_group_size' for write_table() and
-        # keep the rest for ParquetWriter()
-        row_group_size = write_kwargs.pop("row_group_size", None)
+        # Make every incoming batch conform to the final schema *before* writing
+        for idx, t in enumerate(tables):
+            if output_schema and not t.schema.equals(output_schema):
+                t = t.cast(output_schema, safe=False)
+            tables[idx] = t
 
-        write_path = posixpath.join(path, filename)
-        with self.open_output_stream(write_path) as file:
-            with pq.ParquetWriter(file, output_schema, **write_kwargs) as writer:
-                for table in tables:
-                    table = table.cast(output_schema)
-                    writer.write_table(table, row_group_size=row_group_size)
+        row_group_size = write_kwargs.pop("row_group_size", None)
 
-    def _write_partition_files(
-        self,
-        tables: List["pyarrow.Table"],
-        filename: str,
-        output_schema: "pyarrow.Schema",
-        write_kwargs: Dict[str, Any],
-    ) -> None:
-        import pyarrow as pa
-        import pyarrow.compute as pc
-
-        table = concat(tables, promote_types=False)
-        # Create unique combinations of the partition columns
-        partition_col_values: List[Dict[str, Any]] = (
-            table.select(self.partition_cols)
-            .group_by(self.partition_cols)
-            .aggregate([])
-        ).to_pylist()
-        table_fields = [
-            field for field in output_schema if field.name not in self.partition_cols
-        ]
-        non_partition_cols = [f.name for f in table_fields]
-        output_schema = pa.schema(
-            [field for field in output_schema if field.name not in self.partition_cols]
+        ds.write_dataset(
+            data=tables,
+            base_dir=self.path,
+            schema=output_schema,
+            basename_template=f"{filename}-{{i}}.parquet",
+            filesystem=self.filesystem,
+            partitioning=self.partition_cols,
+            format="parquet",
+            existing_data_behavior="overwrite_or_ignore",
+            partitioning_flavor="hive",
+            use_threads=True,
+            min_rows_per_group=row_group_size if row_group_size else 0,
+            max_rows_per_group=row_group_size if row_group_size else 1024 * 1024,
+            file_options=ds.ParquetFileFormat().make_write_options(**write_kwargs),
```

Comment:
I'm assuming you already verified that this properly forwards write_kwargs to the writer, right?